File: sdk/canton/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/ledger/participant/state/ReassignmentCommand.scala
// Copyright (c) 2025 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.digitalasset.canton.ledger.participant.state

import com.digitalasset.canton.data.CantonTimestamp
import com.digitalasset.canton.topology.SynchronizerId
import com.digitalasset.canton.util.ReassignmentTag.{Source, Target}
import com.digitalasset.daml.lf.value.Value

sealed trait ReassignmentCommand {
  def sourceSynchronizer: Source[SynchronizerId]
  def targetSynchronizer: Target[SynchronizerId]
}

object ReassignmentCommand {
  final case class Unassign(
      sourceSynchronizer: Source[SynchronizerId],
      targetSynchronizer: Target[SynchronizerId],
      contractId: Value.ContractId,
  ) extends ReassignmentCommand

  final case class Assign(
      sourceSynchronizer: Source[SynchronizerId],
      targetSynchronizer: Target[SynchronizerId],
      unassignId: CantonTimestamp,
  ) extends ReassignmentCommand
}

File: sdk/canton/community/bindings-java/src/main/java/com/daml/ledger/javaapi/data/ReassignmentCommand.java
// Copyright (c) 2025 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.daml.ledger.javaapi.data;

import com.daml.ledger.api.v2.ReassignmentCommandOuterClass;
import org.checkerframework.checker.nullness.qual.NonNull;

import java.util.Objects;
import java.util.Optional;

import static java.util.Optional.empty;

public abstract class ReassignmentCommand {
  public static ReassignmentCommand fromProtoCommand(
      ReassignmentCommandOuterClass.ReassignmentCommand command) {
    switch (command.getCommandCase()) {
      case ASSIGN_COMMAND:
        return AssignCommand.fromProto(command.getAssignCommand());
      case UNASSIGN_COMMAND:
        return UnassignCommand.fromProto(command.getUnassignCommand());
      case COMMAND_NOT_SET:
      default:
        throw new ProtoReassignmentCommandUnknown(command);
    }
  }

  public ReassignmentCommandOuterClass.ReassignmentCommand toProtoCommand() {
    ReassignmentCommandOuterClass.ReassignmentCommand.Builder builder =
        ReassignmentCommandOuterClass.ReassignmentCommand.newBuilder();
    if (this instanceof UnassignCommand) {
      builder.setUnassignCommand(((UnassignCommand) this).toProto());
    } else if (this instanceof AssignCommand) {
      builder.setAssignCommand(((AssignCommand) this).toProto());
    } else {
      throw new ReassignmentCommandUnknown(this);
    }
    return builder.build();
  }
}

class ReassignmentCommandUnknown extends RuntimeException {
  public ReassignmentCommandUnknown(ReassignmentCommand command) {
    super("Reassignment command unknown " + command.toString());
  }
}

class ProtoReassignmentCommandUnknown extends RuntimeException {
  public ProtoReassignmentCommandUnknown(
      ReassignmentCommandOuterClass.ReassignmentCommand command) {
    super("Reassignment command unknown " + command.toString());
  }
}
